### PR TITLE
Prometheus support parse exemplars from native histogram

### PR DESCRIPTION
**Pull request by fatsheep9146**
following
https://github.com/prometheus/prometheus/pull/13449#pullrequestreview-1844632501

Prometheus support parse exemplars from native histogram.

For now, prometheus will parse the exemplar from the `exemplars` field of the data if the field is not empty, otherwise it will parse like before, which is parsing the exemplar from the `bucket` field.

### CODE REVIEW

**Comment by beorn7** [comment]
I assume this is still WIP? (You said so in the title, but didn't submit it as WIP. And tests are still failing.)

**Comment by beorn7** [comment]
I have converted the PR to a draft. Please mark it as "ready for review" as appropriate.

**Comment by fatsheep9146** [comment]
> I have converted the PR to a draft. Please mark it as "ready for review" as appropriate.

Thanks, this pr is still in progress. @beorn7

File: model/textparse/protobufparse_test.go [hunk 1]
```diff
@@ -595,6 +595,94 @@ metric: <
...
+	    nanos: 146848499
+	  >
+	>
+	exemplars: <
+	  label: <
+	    name: "dummyID"
+	    value: "5617"
+	  >
+	  value: -0.00029
+	>
```
[the remark this answers is not in the document]
Comment (beorn7) [reply model/textparse/protobufparse_test.go:681]: In different news, could you add one more exemplar with a timestamp at the end to test if multiple exemplars are parsed fine and if they are also parsed after an exemplar without a timestamp?

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thank you very much. Looks good code-wise, just some comments about doc comments and tests.

File: model/textparse/protobufparse.go [hunk 2]
```diff
@@ -75,6 +75,8 @@ type ProtobufParser struct {
 
 	// The following are just shenanigans to satisfy the Parser interface.
 	metricBytes *bytes.Buffer // A somewhat fluid representation of the current metric.
+
+	exemaplerPos int
```
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply model/textparse/protobufparse.go]: Done

File: model/textparse/protobufparse.go [hunk 3]
```diff
@@ -317,28 +319,42 @@ func (p *ProtobufParser) Exemplar(ex *exemplar.Exemplar) bool {
 	case dto.MetricType_COUNTER:
 		exProto = m.GetCounter().GetExemplar()
 	case dto.MetricType_HISTOGRAM, dto.MetricType_GAUGE_HISTOGRAM:
-		bb := m.GetHistogram().GetBucket()
 		isClassic := p.state == EntrySeries
-		if p.fieldPos < 0 {
-			if isClassic {
-				return false // At _count or _sum.
+		if !isClassic && len(m.GetHistogram().GetExemplars()) > 0 {
```
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply model/textparse/protobufparse.go:324]: Done

**Comment by beorn7** [comment]
On my list…

File: model/textparse/protobufparse.go [hunk 4]
```diff
@@ -56,6 +56,8 @@ type ProtobufParser struct {
 	fieldPos    int
 	fieldsDone  bool // true if no more fields of a Summary or (legacy) Histogram to be processed.
 	redoClassic bool // true after parsing a native histogram if we need to parse it again as a classic histogram.
+	// exemplarPos is the position within exemplars slice of native exemplars.
```
Comment (beorn7) [inline model/textparse/protobufparse.go:59]:
```suggestion
	// exemplarPos is the position within the exemplars slice of a native histogram.
```

File: model/textparse/protobufparse.go [hunk 5]
```diff
@@ -304,8 +306,9 @@ func (p *ProtobufParser) Metric(l *labels.Labels) string {
 
 // Exemplar writes the exemplar of the current sample into the passed
 // exemplar. It returns if an exemplar exists or not. In case of a native
-// histogram, the legacy bucket section is still used for exemplars. To ingest
-// all exemplars, call the Exemplar method repeatedly until it returns false.
+// histogram, the exemplars in native histogram will be returned.
+// If this field is empty, the legacy bucket section is still used for exemplars.
```
Comment (beorn7) [inline model/textparse/protobufparse.go:310]:
```suggestion
// histogram, the exemplars in the native histogram will be returned.
// If this field is empty, the classic bucket section is still used for exemplars.
```

**Review by beorn7 — APPROVED** [review]
There are a few more word-smithing nits, but I'll commit them directly.

Thank you very much.

**Pull request merged**